crates/core/src/network/io/reactor.rs

```diff
@@ -0,0 +1,465 @@
+        }
+    }
+
+    fn switch_to_draining(&mut self, reason: DrainReason, peer_router: &impl PeerRouting) {
+        trace!("Connection is draining");
+        self.send_drain_signal(reason);
+        self.seen_versions = None;
+        self.state = State::Draining {
+            drain_timeout: Box::pin(tokio::time::sleep(Duration::from_secs(15))),
+        };
+        peer_router.deregister(&self.connection);
+    }
+
+    pub async fn run_reactor<S>(
+        mut self,
+        router: impl Handler,
+        mut incoming: S,
+        conn_tracker: impl ConnectionTracking,
+        peer_router: impl PeerRouting,
+        should_register: bool,
+    ) -> anyhow::Result<()>
+    where
+        S: Stream<Item = Result<Message, ProtocolError>> + Unpin + Send,
+    {
+        let current_task = TaskContext::current();
+        let metadata = Metadata::current();
+        self.seen_versions = Some(MetadataVersions::new(metadata));
+        Span::current().record("task_id", tracing::field::display(current_task.id()));
+        let mut cancellation = std::pin::pin!(current_task.cancellation_token().cancelled());
+
+        if should_register {
+            peer_router.register(&self.connection);
+        }
+        conn_tracker.connection_created(&self.connection);
+
+        loop {
+            let decision = match self.state {
+                State::Active => {
+                    // read a message from the stream
+                    tokio::select! {
+                        biased;
+                        _ = &mut cancellation => {
+                            if TaskCenter::is_shutdown_requested() {
+                                // We want to make the distinction between whether we are terminating the
+                                // connection, or whether the node is shutting down.
+                                Decision::Drain(DrainReason::Shutdown)
+                            } else {
+                                Decision::Drain(DrainReason::ConnectionDrain)
+                            }
+                            // we only drain the connection if we were the initiators of the termination
+                        },
+                        msg = incoming.next() => {
+                            self.handle_message(msg, &router).await
+                        }
+                    }
+                }
+                State::Draining {
+                    ref mut drain_timeout,
+                } => {
+                    tokio::select! {
+                        _ = drain_timeout => {
+                            debug!("Drain timed out, closing connection");
+                            Decision::Drop
+                        },
+                        msg = incoming.next() => {
+                            self.handle_message(msg, &router).await
+                        }
+                    }
+                }
+                State::WaitForEgress => {
+                    self.tx.take();
```

Comment:
I think `tx` should already be dropped (when transitioning into the `WaitForEgress` state).

Reply:
Ye, just being extra defensive.